flashlight/meter/PerformanceAudit.h

```diff
@@ -0,0 +1,353 @@
+  }
+
+  /**
+   * @return returns the mode preferred computation mode for the corresponding
+   * kernel.
+   */
+  inline KernelMode mode() const {
+    return mode_;
+  }
+
+  inline bool algoMustSupportF16() {
+    return (useMixedPrecision_ && mode_ != KernelMode::F32);
+  }
+
+  /**
+   * Audits the performance of a given convolution kernel and measures it's
+   * execution time for different arithmetic precision to find the computation
+   * settings for which the kernel yields the highest performance.
+   *
+   * @param[in] kernel the target convolution kernel for which this component
+   * will measure the performance.
+   * @param[in] offset initial offset to add to measurements.
+   */
+  __forceinline void audit(std::function<void()> kernel, double offset = 0) {
+    start(offset);
+    kernel();
+    stop();
+  }
+
+  inline void reset() {
+    mode_ = KernelMode::F32;
+    auditDone_ = false;
+
+    countF32_ = 0;
+    countF32C_ = 0;
+    countF16_ = 0;
+
+    timeF32_ = 0;
+    timeF32C_ = 0;
+    timeF16_ = 0;
+  }
+};
+
+using KernelPerfAuditPtr = std::shared_ptr<KernelPerfAudit>;
+
+/**
+ * Set of KernelPerfAudit objects for different CUDA kernels of each
+ * convolution.
+ */
+struct ConvPerfAudit {
+ private:
+  FL_SAVE_LOAD(bwdBias, bwdData, bwdFilter, useMixedPrecision_)
+
+ public:
+  KernelPerfAuditPtr bwdBias;
+  KernelPerfAuditPtr bwdData;
+  KernelPerfAuditPtr bwdFilter;
+  bool useMixedPrecision_;
+
+  /**
+   * @param[in] useMixedPrecision determines if the convolution kernel is
+   * allowed to use mixed precision. If yes, we might use it depending on the
+   * performance benefit.
+   */
+  ConvPerfAudit() {
+    bwdBias = std::make_shared<KernelPerfAudit>(false);
+    bwdData = std::make_shared<KernelPerfAudit>(false);
+    bwdFilter = std::make_shared<KernelPerfAudit>(false);
+  }
+
+  inline ConvolutionMode getConvolutionMode() const {
+    if (useMixedPrecision_) {
+      if (!(bwdBias->prefF32() || bwdData->prefF32() || bwdFilter->prefF32())) {
+        return ConvolutionMode::DEFAULT;
+      }
+      if (bwdBias->prefF32() && bwdData->prefF32() && bwdFilter->prefF32()) {
+        return ConvolutionMode::FORCED_F32;
+      }
+      return ConvolutionMode::HYBRID;
+    }
+    return ConvolutionMode::DEFAULT;
+  }
+
+  inline bool auditDone() const {
+    return (
+        bwdBias->auditDone() && bwdData->auditDone() && bwdFilter->auditDone());
+  }
+
+  inline void reset() const {
+    bwdBias->reset();
+    bwdData->reset();
+    bwdFilter->reset();
+  }
+
+  inline bool setUseMixedPrecision(bool useMixedPrecision) {
+    useMixedPrecision_ = useMixedPrecision;
+    bwdBias->setUseMixedPrecision(useMixedPrecision);
+    bwdData->setUseMixedPrecision(useMixedPrecision);
+    bwdFilter->setUseMixedPrecision(useMixedPrecision);
+  }
+};
+
+using ConvPerfAuditPtr = std::shared_ptr<ConvPerfAudit>;
+
+/**
+ * @return returns a pointer to a ConvPerfAudit struct.
+ */
+inline ConvPerfAuditPtr getConvAuditHandle() {
```

Comment:
Still not crazy about the perf audit, but starting to warm up to it a bit. Going to give more thought to how we might want to add this and will give feedback shortly afterwards.

Reply:
Sure. Feel free to share your thoughts on how you'd prefer to change it :).

Reply:
See my comment below.